Runtime/Haptics/HapticPulser.cs

```diff
@@ -14,19 +13,17 @@ public abstract class HapticPulser : HapticProcess
         /// <summary>
         /// The intensity of the haptic rumble.
         /// </summary>
-        [Serialized, /*Validated*/]
-        [field: Range(0f, 1f), DocumentedByXml]
+        [Serialized,]
+        [field: DocumentedByXml, Range(0f, 1f)]
         public float Intensity { get; set; } = 1f;
 
         /// <summary>
-        /// Handles changes to <see cref="Intensity"/>.
+        /// Called after <see cref="Intensity"/> has been changed.
         /// </summary>
-        /// <param name="previousValue">The previous value.</param>
-        /// <param name="newValue">The new value.</param>
-        /*[CalledBySetter(nameof(Intensity))]*/
-        protected virtual void OnIntensityChange(float previousValue, ref float newValue)
+        [CalledAfterChangeOf(nameof(Intensity))]
+        protected virtual void OnAfterIntensityChange()
         {
-            newValue = Mathf.Clamp01(newValue);
+            Intensity = Mathf.Clamp01(Intensity);
```

Comment:
This loops infinitely - check if the desired result `Mathf.Clamp01(Intensity)` isn't already the value of `Intensity`.

Reply:
what's the most efficient check here?

```
float clampedIntensity = Mathf.Clamp01(Intensity);
if (! Intensity.ApproxEquals(clampedIntensity)) {
  Intensity = clampedIntensity;
}
```

or just simply

```
if (Intensity < 0 || Intensity > 1)
{
    Intensity = Mathf.Clamp01(Intensity);
}
```

Reply:
Let's use the former (without the space after the negation 😉) - the difference is negligible and that first solution reads easier as it says "this is what it should be, is it already? OK set it".

Reply:
The former always performs a clamp operation even when it's not required. to me the latter is easier to understand.

if the number is less than 0 or greater than 1 then clamp between 0 and 1

you also don't have to do the clamp every time.

Reply:
The most efficient solution is to clamp yourself, as you're already doing half of the job the Clamp call does by doing the if check here. But that's way too long.

Go with whatever you feel like.